doc/ircd.8
.\" @(#)ircd.8 7b10 25 Oct 2001
.TH IRCD 8 "2021-01-01" "ircd-hybrid"
.SH NAME
ircd \- ircd-hybrid: an advanced, lightweight Internet Relay Chat Daemon
.SH SYNOPSIS
.hy 0
.IP \fBircd\fP
[-configfile filename] [-klinefile filename] [-dlinefile filename]
[-xlinefile filename] [-resvfile filename] [-logfile filename] [-pidfile filename]
[-foreground] [-version] [-help]
.SH DESCRIPTION
.LP
\fIircd\fP is the server (daemon) program for the Internet Relay Chat
Program. The \fIircd\fP is a server in that its function is to "serve"
the client program \fIirc(1)\fP and other compatible programs with
messages and commands. All commands and user messages are passed
directly to the \fIircd\fP for processing and relaying to other IRC
servers.

This is \fIircd-hybrid\fP, the highly modified (and hopefully improved)
variant of the original \fIircd\fP program by Jarkko Oikarinen.
.SH OPTIONS
.TP
.B \-configfile filename
Specifies the ircd.conf file to be used for this \fIircd\fP. The option
is used to override the default ircd.conf given at compile time. By default,
this is "etc/ircd.conf" within the prefix you installed the \fIircd\fP in.
.TP
.B \-klinefile filename
.TP
.B \-dlinefile filename
.TP
.B \-xlinefile filename
.TP
.B \-resvfile filename
Specifies the database files to be used for D-lines (host bans), K-lines (hostmask bans),
X-lines (gecos bans) and RESVs (jupes) which by default are within the var/lib/
directory of your installation prefix as dline.json, kline.json, xline.json, and resv.json.
.TP
.B \-logfile filename
Specifies an alternative logfile to be used than that specified in include/defaults.h
.TP
.B \-pidfile filename
Specifies the file used by the \fIircd\fP to store its process ID. The option is
used to override the default ircd.pid given at compile time.
.TP
.B \-foreground
Makes \fIircd\fP run in the foreground.
.TP
.B \-version
Makes \fIircd\fP print its version and exit.
.TP
.B \-help
Makes \fIircd\fP print available options and exit.
.SH USAGE
If you plan to connect your IRC server to an existing IRC network,
you will need to alter your local IRC configuration file (typically named
"ircd.conf") so that it will accept and make connections to other IRC servers.
This file contains the hostnames, IP addresses and sometimes passwords
for connections to other \fIircds\fP around the world.

The example reference.ircd.conf in the doc/ directory documents itself. Read it
carefully or you may expose a risk on your network simply by having your
\fIircd\fP running!
.LP
.SH BOOTING THE SERVER
It is sufficient to type \fIircd\fP at the command line to start
\fIircd-hybrid\fP into the background. If you wish to run \fIircd\fP in the
foreground (perhaps for debugging purposes), use the \fB-foreground\fP
parameter.

The \fIircd-hybrid\fP package in your distribution may ship with an rc-script
which handles this for you. In Debian, it is \fB/etc/init.d/ircd-hybrid\fP.
.SH EXAMPLE
# ircd -configfile /usr/share/ircd/ircd.conf -logfile /var/log/ircd.log

Starts \fIircd\fP with the config file /usr/share/ircd/ircd.conf and
with a log file /var/log/ircd.log. It implicitly forks back to the
console -- to prevent this, use \fB-foreground\fP.

Many of the above paths are hard-coded at compile time in the
\fBinclude/defaults.h\fP file, so it may be convenient to edit these to
suit your needs before you compile the \fIircd\fP.
.SH COPYRIGHT
(c) 1988,1989 University of Oulu, Computing Center, Finland,
.LP
(c) 1988,1989 Department of Information Processing Science,
University of Oulu, Finland
.LP
(c) 1988-1991 Jarkko Oikarinen
.LP
(c) 1997-2026 ircd-hybrid development team
.LP
For full COPYRIGHT, see the COPYING file within the IRC source package.
.LP
.RE
.SH BUGS
If you find a bug and you have a core file from \fIircd\fP as a result, use a
debugger like \fBgdb(1)\fP to process the core file and send a backtrace
to \fBbugs@ircd-hybrid.org\fP.
.SH AUTHORS
irc2.8 and earlier: Jarkko Oikarinen, currently jto@tolsun.oulu.fi.
.LP
ircd-hybrid: ircd-hybrid development team, bugs@ircd-hybrid.org.
.LP
Manual page written by Jeff Trim, jtrim@orion.cair.du.edu,
later modified by jto@tolsun.oulu.fi.
.LP
Updated by W. Campbell, wcampbel@botbay.net, Edward Brocklesby,
ejb@klamath.uucp.leguin.org.uk, and highly revised by Joshua Kwan
(joshk@triplehelix.org) for the Debian distribution.


NEWS
-- Noteworthy changes in version 8.2.40 (2021-??-??)
* `channel::disable_fake_channels` now also blocks strikethrough text (0x1E (30))
* `KNOCK` can now be used on private channels. Channel mode +K may be used instead
  to control whether or not `KNOCK` is allowed on a channel.
* Support for /dev/poll has been dropped
* `INFO` now also reports configured values of `channel::default_join_flood_count`
  and `channel::default_join_flood_time`
* The founder (+q) and admin (+a) channel prefix modes have been implemented.
  These are optional and can be enabled with the new `channel::enable_owner`
  and `channel::enable_admin` configuration directives.


-- Noteworthy changes in version 8.2.39 (2021-08-14)
* `hidechans` and `hideidle` may now be set in `general::oper_only_umodes`
* OpenSSL 3.0 compatibility improvements
* Fixed crashing issue in the mkpasswd tool with older glibc
* Fixed issue where the ircd wouldn't listen on both IPv4 and IPv6 ports in
  case only a port but no address has been specified in a listen {} block
* Fixed a `STATS P` displaying issue with compressed ipv6 loopback/inaddr_any


-- Noteworthy changes in version 8.2.38 (2021-01-17)
* Backwards compatibility mode for ircd-hybrid 8.2.22 and below has been removed
* Minimum supported version of the Anope IRC services is now 2.0.8
* Fixed displaying issue with channel `MODE` messages
* Characters valid for user names (ident) for local clients are now `[A-Za-z0-9._-]`.
  Special characters like ``}|{`^]\[$``, that were allowed previously, are no longer
  valid. Also user names may not begin with `.`, `_`, or `-`.
* The `general::dots_in_ident` configuration directive has been renamed to
  `general::specials_in_ident` and now covers the `-` and `_` characters in
  addition to `.`
* Fixed issue with clients connected from IPv6 localhost that, depending on the
  system's `getaddrinfo` implementation, could lead to a crash of remote servers


-- Noteworthy changes in version 8.2.37 (2020-12-26)
* Implemented IRCv3 `CAP 302`
* Implemented IRCv3 `cap-notify` capability
* In case a client sends multiple `PASS` messages, ircd only accepts the password
  sent in the first `PASS` message instead of the last one
* Missing help file for `MONITOR` has been added
* Fixed slightly off memory count for motds in `STATS z`


-- Noteworthy changes in version 8.2.36 (2020-12-04)
* The old `WATCH` implementation has been replaced with IRCv3.2 `MONITOR`. For this,
  the `general::max_watch` configuration directive has been renamed to
  `general::max_monitor`.
* User mode `B` has been implemented. Clients with that mode set are marked as a
  bot in both `WHOIS` and `WHO`. This mode can only be set by IRC operators (as
  long as the `bot` directive is set in `general::oper_only_umodes`), servers,
  and services.
  This mode can for example be used to mark HOPM as official network bot.


-- Noteworthy changes in version 8.2.35 (2020-11-14)
* Fixed issue where servers could propagate truncated ban masks during net-join


-- Noteworthy changes in version 8.2.34 (2020-11-01)
* Re-implement backwards compatibility mode for old ircd-hybrid 8.2.23 and below.
  Final removal is scheduled for early 2021.
* For consistency, the `general::network_desc` configuration directive has been
  renamed to `general::network_description`
* Server notices that previously have been sent to user modes +f and +u are now
  sent to user mode +j (reject notices) instead. Due to this change, `unauth` and
  `full` have been removed from `general::oper_only_umodes`, `general::oper_umodes`,
  and `operator::umodes`.
* User mode +b (bots) has been replaced with +f (flood). Spam/flood notices are now
  sent to this user mode. For this, `bots` has been removed from
  `general::oper_only_umodes`, `general::oper_umodes`, and `operator::umodes`.
  `flood` has been added instead to these configuration directives.
* Stricten server/user ID validation. IDs have to be all uppercase everywhere now.


-- Noteworthy changes in version 8.2.33 (2020-09-07)
* Added `client` option to `listener::flags`
* Added `defer` option to `listener::flags`
* IRC operators may now use CIDR notation in `WHO`


-- Noteworthy changes in version 8.2.32 (2020-08-16)
* `RESV` couldn't be issued in case no reason has been supplied. This has been fixed.
* Fixed possible `RPL_WHOISCHANNELS` line truncation of remote replies
* Extban $t of type `matching` has been implemented. This extban allows matching
  based on TLS protocol version and/or cipher suite
* Implemented channel mode `K`. `KNOCK` cannot be used on channels with that mode set
* `STATS ?` is now oper-only


-- Noteworthy changes in version 8.2.31 (2020-05-03)
* macOS compatibility fixes
* Removed `spoof_notice` from `auth::flags`. With vhosts now this notice doesn't make
  much sense anymore
* Fixed issue where ban masks might become malformed if set by remote clients/servers
* Fixed issue with channel mode +c where high ascii characters can be erroneously
  detected as control characters


-- Noteworthy changes in version 8.2.30 (2020-03-01)
* `QUOTE HELP extban` didn't work because the extban help file hasn't been
  installed properly during `make install`
* Fixed broken libcrypto detection which caused the ircd not to work with
  OpenSSL under certain circumstances
* Extban $n of type `acting` has been implemented. This extban prevents
  matching users from changing their nick while in the channel. Users
  with voice or above are not affected.
* Channel mode +N has been changed so channel members with +v can change
  their nick name as well


-- Noteworthy changes in version 8.2.29 (2020-02-19)
* Extbans have been implemented. Currently supported extbans:

  Matching:

  * `$a:<account>`   Matches users logged into a matching account.
  * `$c:<channel>`   Matches users that are on the given channel. An additional
                     prefix of either @, %, or + can be specified to test for
                     certain channel privileges.

src/ircd.c
/*
 *  ircd-hybrid: an advanced, lightweight Internet Relay Chat Daemon (ircd)
 *
 *  Copyright (c) 1997-2026 ircd-hybrid development team
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301
 *  USA
 */

/**
 * @file ircd.c
 * @brief Starts up and runs the ircd.
 *
 * This file contains the main entry point and functionality for starting up
 * and running the ircd (Internet Relay Chat Daemon). It initializes various
 * components, sets up event handlers, and manages server configurations.
 */

#include <jansson.h>  /**< Jansson JSON library. */

#include <fcntl.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/types.h>

#include "comm.h"
#include "event.h"
#include "fdlist.h"
#include "io_getopt.h"
#include "io_pidfile.h"
#include "io_rlimit.h"
#include "io_string.h"
#include "io_time.h"
#include "list.h"
#include "log.h"
#include "memory.h"
#include "module.h"
#include "res.h"
#include "rng_mt.h"

#include "channel_mode.h"
#include "client.h"
#include "cloak.h"
#include "conf.h"
#include "conf_class.h"
#include "conf_db.h"
#include "defaults.h"
#include "extban.h"
#include "hash.h"
#include "id.h"
#include "ipcache.h"
#include "ircd.h"
#include "ircd_exit.h"
#include "ircd_hook.h"
#include "ircd_signal.h"
#include "links_cache.h"
#include "motd.h"
#include "send.h"
#include "server.h"
#include "server_capab.h"

event_manager_t ircd_event_manager;

/**
 * @struct SetOptions
 * @brief Structure for server configuration options.
 *
 * The `SetOptions` structure holds various options related to server configuration,
 * allowing modification through the SET irc command. It includes flags and settings
 * for features such as auto connection, maximum clients, and flood control.
 */
struct SetOptions GlobalSetOptions;

/**
 * @struct Counter
 * @brief Structure holding various server activity counters.
 *
 * The `Counter` structure tracks various server activity metrics, including total
 * client connections, IRC operator count, invisible clients, and peak client counts.
 */
struct Counter Count;

/**
 * @struct ServerState_t
 * @brief Structure representing the server's foreground state.
 *
 * The `ServerState_t` structure holds a flag indicating whether the server should
 * run in the foreground. The flag is set based on the command line parameter
 * '-foreground' when launching the ircd process.
 */
struct ServerState_t server_state;

/**
 * @struct ServerStatistics
 * @brief Structure for server communication and connection statistics.
 *
 * The `ServerStatistics` structure contains statistics related to server communication
 * and connections. These metrics provide insights into data transmission, connection
 * durations, and various activities between clients and servers.
 */
struct ServerStatistics ServerStats;

/**
 * @struct Connection
 * @brief Structure representing the connection details for this server.
 *
 * The `Connection` structure encapsulates the connection details for the server,
 * including information such as connection status, socket details, and connection settings.
 */
static struct Connection meConnection;

/**
 * @struct Client
 * @brief Structure representing the client details for this server.
 *
 * The `Client` structure represents the client details for the server, with a connection
 * link pointing to the associated `Connection` structure. This structure provides a
 * comprehensive view of the client's attributes within the context of the server.
 */
struct Client me = { .connection = &meConnection };

/**
 * @var char **myargv
 * @brief Pointer to the command line arguments.
 */
char **myargv;

/**
 * @var const char *logFileName
 * @brief Pointer to the filename for the ircd.log file.
 */
const char *logFileName = LPATH;

/**
 * @var const char *pidFileName
 * @brief Pointer to the filename for the process ID file.
 */
const char *pidFileName = PPATH;

/**
 * @var bool printVersion
 * @brief Flag indicating whether to print the version and exit.
 */
static bool printVersion;

/**
 * @var struct io_getopt myopts[]
 * @brief Array of command-line options and their descriptions.
 */
static struct io_getopt myopts[] =
{
  { "configfile", 'c', &ConfigGeneral.configfile, STRING, "File to use for ircd.conf" },
  { "klinefile", 'k', &ConfigGeneral.klinefile, STRING, "File to use for kline database" },
  { "dlinefile", 'd', &ConfigGeneral.dlinefile, STRING, "File to use for dline database" },
  { "xlinefile", 'x', &ConfigGeneral.xlinefile, STRING, "File to use for xline database" },
  { "resvfile", 'r', &ConfigGeneral.resvfile, STRING, "File to use for resv database" },
  { "logfile", 'l', &logFileName, STRING, "File to use for ircd.log" },
  { "pidfile", 'p', &pidFileName, STRING, "File to use for process ID" },
  { "foreground", 'f', &server_state.foreground, BOOLEAN, "Run in foreground (don't detach)" },
  { "version", 'v', &printVersion, BOOLEAN, "Print version and exit" },
  { "help", 'h', NULL, USAGE, "Print this text" },
  { NULL, 0, NULL, STRING, NULL }
};

/**
 * @brief Main IO loop for processing events and managing server activities.
 *
 * This function runs continuously, handling various tasks such as
 * processing events, checking for rehash signals, and managing client connections.
 */
static void
io_loop(void)
{
  while (true)
  {
    comm_select(comm_get_select_timeout(ircd_event_manager));
    exit_aborted_clients();
    free_exited_clients();

    io_time_update_cache();
    event_run(ircd_event_manager);

    /* Check to see whether we have to rehash the configuration. */
    if (dorehash)
    {
      conf_rehash(true);
      dorehash = 0;
    }

    if (doremotd)
    {
      motd_recache();
      sendto_clients(UMODE_SERVNOTICE, SEND_RECIPIENT_OPER_ALL, SEND_TYPE_NOTICE,
                     "Got signal SIGUSR1, reloading motd file(s)");
      doremotd = 0;
    }
  }
}

/**
 * @brief Initializes global server configuration options.
 *
 * This function sets default values for various global server configuration options
 * needed during runtime. It is responsible for initializing the `GlobalSetOptions`
 * structure, which holds key settings that can be modified using the SET irc command.
 */
static void
initialize_global_set_options(void)
{
  GlobalSetOptions.maxclients = ConfigServerInfo.default_max_clients;
  GlobalSetOptions.autoconnect = true;
  GlobalSetOptions.spam_time = MIN_JOIN_LEAVE_TIME;
  GlobalSetOptions.spam_num = MAX_JOIN_LEAVE_COUNT;
  GlobalSetOptions.floodcount = ConfigGeneral.default_floodcount;
  GlobalSetOptions.floodtime = ConfigGeneral.default_floodtime;
  GlobalSetOptions.joinfloodcount = ConfigChannel.default_join_flood_count;
  GlobalSetOptions.joinfloodtime = ConfigChannel.default_join_flood_time;
}

/**
 * @brief Handles out-of-memory conditions.
 *
 * This function reports an out-of-memory condition and restarts the program.
 */
static void
ircd_oom(void)
{
  ircd_exit(IRCD_EXIT_RESTART, "out of memory");
}

/**
 * @brief Handles time provider API error conditions.
 *
 * This function reports an error condition encountered in the time provider API using the provided
 * error message. It exits with EXIT_FAILURE if the error code is IO_TIME_ERR_INIT, otherwise, it restarts the program.
 *
 * @param error_code The error code indicating the type of error in the time provider API.
 * @param message The error message providing details about the time provider API error.
 */
static void
ircd_time_failure(enum io_time_error_code error_code, const char *message)
{
  if (error_code == IO_TIME_ERR_INIT)
  {
    fprintf(stderr, "ERROR: %s\n", message);
    exit(EXIT_FAILURE);  /* Exit with failure if initialization error. */
  }

  ircd_exit(IRCD_EXIT_RESTART, message);  /* Restart the program for other errors. */
}

/**
 * @brief Prints startup information including version and process ID.
 *
 * This function prints information about the server version, process ID,
 * and whether it is running in the background or foreground.
 *
 * @param pid Process ID of the ircd server.
 */
static void
print_startup(pid_t pid)
{
  printf("ircd: version %s\n", IRCD_VERSION);
  printf("ircd: pid %d\n", pid);
  printf("ircd: running in %s mode from %s\n",
         server_state.foreground ? "foreground": "background", ConfigGeneral.dpath);
}

/**
 * @brief Transforms the current process into a daemon for background execution.
 *
 * This function creates a child process using fork(), allowing the parent process
 * to exit. The child becomes a daemon by detaching from the terminal and setting
 * up a new session using setsid(). Standard input, output, and error are redirected
 * to /dev/null to isolate the daemon from the terminal.
 */
static void
make_daemon(void)
{
  pid_t pid = fork();
  if (pid == -1)
  {
    perror("fork");
    exit(EXIT_FAILURE);
  }
  else if (pid > 0)
  {
    print_startup(pid);
    exit(EXIT_SUCCESS);
  }

  if (setsid() == -1)
  {
    perror("setsid");
    exit(EXIT_FAILURE);
  }

  /* Connect stdin, stdout, and stderr to /dev/null */
  const int null_fd = open("/dev/null", O_RDWR);
  if (null_fd == -1)
  {
    perror("failed to open /dev/null");
    exit(EXIT_FAILURE);
  }

  if (dup2(null_fd, STDIN_FILENO ) == -1 ||
      dup2(null_fd, STDOUT_FILENO) == -1 ||
      dup2(null_fd, STDERR_FILENO) == -1)
  {
    perror("failed to redirect standard fds");
    exit(EXIT_FAILURE);
  }

  if (null_fd > STDERR_FILENO)
    close(null_fd);
}

/**
 * @brief Main function to initialize and run the IRC server.
 *
 * This is the main entry point for the ircd server. It initializes various
 * components, sets up signal handling, reads configuration files, and runs
 * the main IO loop to handle server activities.
 *
 * @param argc Number of command line arguments.
 * @param argv Array of command line argument strings.
 * @return Returns 0 upon successful execution.
 */
int
main(int argc, char *argv[])
{
  /* Check to see if the user is running us as root, which is a nono */
  if (geteuid() == 0)
  {
    fprintf(stderr, "ERROR: This server won't run as root/superuser\n");
    return -1;
  }

  io_set_oom_handler(ircd_oom);

  io_rlimit_set_max_core();

  io_rlimit_set_max_nofile();

  io_time_set_error_callback(ircd_time_failure);
  io_time_init();

  /* Configure libjansson to use io_calloc and io_free. */
  json_set_alloc_funcs(io_calloc, io_free);

  /*
   * Calculate the seed using multiple sources of entropy
   * - Real-time and monotonic clocks provide high-resolution time values
   * - Process ID adds process-specific uniqueness
   * This combination aims to create a seed that is harder to predict
   */
  const uint32_t seed = (uint32_t)(
     (io_time_get(IO_TIME_REALTIME_SEC) % UINT32_MAX) ^  /* Real-time seconds (modulo to fit in 32 bits) */
      io_time_get(IO_TIME_REALTIME_NSEC) ^  /* Real-time nanoseconds */
    ((io_time_get(IO_TIME_MONOTONIC_SEC) % UINT32_MAX) ^  /* Monotonic seconds (modulo to fit in 32 bits) */
      io_time_get(IO_TIME_MONOTONIC_NSEC) ^  /* Monotonic nanoseconds */
    ((uint32_t)(getpid() & 0xFFFF) << 16))  /* Masked and shifted PID for added entropy */
  );

  init_genrand(seed);

  ConfigGeneral.dpath      = DPATH;
  ConfigGeneral.spath      = SPATH;
  ConfigGeneral.configfile = CPATH;    /* Server configuration file */
  ConfigGeneral.klinefile  = KPATH;    /* Server kline file         */
  ConfigGeneral.xlinefile  = XPATH;    /* Server xline file         */
  ConfigGeneral.dlinefile  = DLPATH;   /* dline file                */
  ConfigGeneral.resvfile   = RESVPATH; /* resv file                 */

  myargv = argv;
  umask(077);  /* umask 077: u=rwx,g=,o= */

  io_getopt(&argc, &argv, myopts);

  if (printVersion)
  {
    printf("ircd: version %s\n", IRCD_VERSION);
    exit(EXIT_SUCCESS);
  }

  if (chdir(ConfigGeneral.dpath))
  {
    perror("chdir");
    exit(EXIT_FAILURE);
  }

  if (server_state.foreground)
    print_startup(getpid());
  else
    make_daemon();

  ircd_signal_init();

  ircd_event_manager = event_manager_create(NULL);
  if (ircd_event_manager == NULL)
  {
    exit(EXIT_FAILURE);
  }

  /* We need this to initialise the fd array before anything else */
  fdlist_init();
  log_add(LOG_TYPE_IRCD, true, 0, logFileName);

  comm_init(ircd_event_manager);  /* This needs to be setup early ! -- adrian */
  tls_init();

  /* Check if there is pidfile and daemon already running */
  if (io_pidfile_create(pidFileName))
    exit(EXIT_FAILURE);

  ircd_hook_init();
  class_init();
  cloak_init();
  module_init();
  conf_read_files(true);   /* cold start init conf files */

  resolver_init(ircd_event_manager);
  ipcache_init();
  channel_mode_init();
  extban_init();
  capab_init();  /* Set up default_server_capabs */
  initialize_global_set_options();  /* Has to be called after conf_read_files() */
  links_cache_init();

  if (string_is_empty(ConfigServerInfo.name))
  {
    log_write(LOG_TYPE_IRCD, "ERROR: No server name specified in serverinfo block.");
    exit(EXIT_FAILURE);
  }

  strlcpy(me.name, ConfigServerInfo.name, sizeof(me.name));

  /* serverinfo {} description must exist.  If not, error out.*/
  if (string_is_empty(ConfigServerInfo.description))
  {
    log_write(LOG_TYPE_IRCD, "ERROR: No server description specified in serverinfo block.");
    exit(EXIT_FAILURE);
  }

  strlcpy(me.info, ConfigServerInfo.description, sizeof(me.info));

  if (string_is_empty(ConfigServerInfo.sid))
  {
    log_write(LOG_TYPE_IRCD, "Generating server ID");
    generate_sid();
    ConfigServerInfo.sid = io_strdup(me.id);
  }
  else
    strlcpy(me.id, ConfigServerInfo.sid, sizeof(me.id));

  init_uid();

  me.uplink = &me;
  me.nexthop = &me;
  me.connection->created_real = io_time_get(IO_TIME_REALTIME_SEC);
  me.connection->created_monotonic = io_time_get(IO_TIME_MONOTONIC_SEC);

  SetMe(&me);
  server_create(&me);

  hash_add_id(&me);
  hash_add_client(&me);

  list_add(&me, &me.global_node, &global_server_list);

  load_kline_database(ConfigGeneral.klinefile);
  load_dline_database(ConfigGeneral.dlinefile);
  load_xline_database(ConfigGeneral.xlinefile);
  load_resv_database(ConfigGeneral.resvfile);

  module_load_all(NULL);

  event_handle_t event_cleanup_tklines = event_create(ircd_event_manager, "cleanup_tklines", cleanup_tklines, 30000, false, NULL, NULL);
  event_schedule_fuzzed(event_cleanup_tklines);

  event_handle_t event_server_connect_auto = event_create(ircd_event_manager, "server_connect_auto", server_connect_auto, 15000, false, NULL, NULL);
  event_schedule_fuzzed(event_server_connect_auto);

  event_handle_t event_save_all_databases = event_create(ircd_event_manager, "save_all_databases", save_all_databases, 300000, false, NULL, NULL);
  event_schedule_fuzzed(event_save_all_databases);

  log_write(LOG_TYPE_IRCD, "Server ready. Running version: %s", IRCD_VERSION);
  io_loop();

  return 0;
}
